### ISSUE TITLE
Ambiguous docs+confusing behavior of softmax for 3+ dimensions

### ISSUE THREAD

**Issue by jekbradbury** [pytorch/pytorch#1020]
The docs for `nn.Softmax`, which are essentially copied over from Torch7's `nn.Softmax`, say "Applies the SoftMax function to an n-dimensional input Tensor, rescaling them so that the elements of the n-dimensional output Tensor lie in the range (0, 1) and sum to 1." This is ambiguous with respect to what exactly should sum to 1 (what dimension/s the softmax should be performed over). What in fact happens, also a holdover from Torch7, is:
- For a 1D input, the softmax takes place over dimension 0 (this was needed in Torch7 to support inputs without a batch dimension, which are not allowed in PyTorch). This is still reasonable default behavior, because there's no reason to use `nn.Softmax` when you have a batch of scalars.
- For a 2D input, the softmax takes place over dimension 1. This is the most common case, and the desired behavior.
- For a 3D input, the softmax takes place over dimension 0. This is usually wrong in PyTorch, since dimension 0 is usually the batch dimension. It may, however, be the right thing to do if dimension 0 is the timestep dimension of a timestep x batch x feature tensor; this is the desired behavior in RNNs with attention.
- For a 4D input, the softmax again takes place over dimension 1.
- For a 5D+ input, THNN gives a RuntimeError.

If all of this is intentional, then we should document it. The ideal solution is probably to keep this behavior and document it but add an optional `dim` argument, as in TensorFlow, that lets the user pick a dimension to softmax over. For reference, the default behavior in TF is to softmax over the last dimension, while the default behavior in Chainer (which can also IIRC be overridden by the user as of v2) is to softmax over dimension 1. It is occasionally also useful to be able to softmax over multiple dimensions at once, but this is easy to emulate with `.view()` so it's not necessary to have in core.
It may also be useful to add a `torch.softmax` with the same behavior as `F.softmax`, to apply the softmax operation to Tensors.

**Comment by RongzhaoZhang** [comment]
I encounter the same problem. But in my version of pytorch (ubuntu16.04+py2.7+anaconda), `nn.Softmax` only allows 2D tensor, while `nn.LogSoftmax` 's behavior is exactly the same.
I like the idea that adds an optional `dim` argument, that would be much more convenient.

**Comment by leihamilton** [comment]
I have the same issue, I have an implementation in torch, that the input is a 4D tensor. But now I am trying to implement in pyTorch, and it gives me an error saying it only allows a 2D tensor. What else can I do to solve this problem before it gets fixed?

**Comment by jekbradbury** [comment]
Chainer just implemented a dim argument this week https://github.com/pfnet/chainer/pull/2538/files

**Comment by pavanramkumar** [comment]
I'm writing a `Function` for this, but am running into errors where neither `Variable` nor `FloatTensor` are acceptable inputs to `torch.nn.Softmax()`.

```python
class MySoftmax(Function):

    def forward(self, input_):
        self.save_for_backward(input_)
        output_ = Variable(torch.zeros(batch_size))
        for sample in range(batch_size):
            output_[sample, :, :] = torch.nn.Softmax()(input_[sample, :, :])
        return output_
    
    def backward(self, grad_output):
        input_ = self.saved_tensors
        grad_input_ = None
        return grad_input
```

I've looked at looping logic but couldn't find something obvious. What am I missing? Thanks!

**Comment by jekbradbury** [comment]
If you want to wrap `nn.Softmax` with additional logic to make it work for different dimensions, you don't have to do this in a `Function` (in fact it's not a good idea to call `Function`s inside other `Function`s). You should just do it in a `Module`:
```python
class MySoftmax(Module):

    def forward(self, input_):
        batch_size = input_.size()[0]
        output_ = torch.cat([F.softmax(input_[i]) for i in range(batch_size)], 0)
        return output_
```

**Comment by pavanramkumar** [comment]
@jekbradbury thanks for this!

> (in fact it's not a good idea to call `Functions` inside other `Functions`)

the above warning could be useful in the [Extending torch.nn](http://pytorch.org/docs/notes/extending.html#extending-torch-nn) page.

**Comment by apaszke** [comment]
Actually after merging the autograd refactor we might enable a format that allows calling other functions inside the functions, but in this case it's not very useful anyway. BTW, it might be faster to transpose all dimensions to a format that's accepted by the backend right now and do an inverse transpose them after computing it.

**Comment by apaszke** [comment]
I'll try to write up an implementation for that later today + maybe will manage to rewrite the kernels to support it out of the box.

**Comment by colesbury** [comment]
The current softmax behavior for choice of dimension is not consistent and doesn't make much sense. We should definitely allow the user to specify the dimension. 

I think we should also choose a default dimension that's consistent across the number of dimensions. That leaves dim 0 and dim -1 (and dim 0 is not what anyone wants). -1 is nice because it treats all the left-most dimensions like a "batch". This is consistent with broadcasting behavior, where the left-most dimensions get expanded. For example:

```F.softmax((input @ weight) + bias)```

The `input @ weight` generalizes from 2d to 3d+ dimensions by treating all the left-most dimensions the same way (as part of a batch/stack). Softmax should do the same thing.

**Comment by maksymbevza** [comment]
Was it fixed?

**Comment by soumith** [comment]
yes, now logsoftmax and softmax take a dimension argument and the docs specify the default argument.

**Comment by schmmd** [comment]
@soumith where in the docs specifies the default argument value?  I'm running into this while upgrading github.com/allenai/allennlp to use PyTorch 0.3.

**Comment by apaszke** [comment]
There's no default. It should be specified explicitly.

**Comment by schmmd** [comment]
I was responding to "the docs specify the default argument".  While it should be specified explicitly, and I get deprecation errors if I do not, I am able to call those methods without an explicit `dims` argument in PyTorch 0.3.
